301-azure-firewall-dns-proxy/README.md

```diff
@@ -0,0 +1,794 @@
+
+![Route Table](images/dns-proxy-without-custom-dns-forwarder.png)
+
+This is the list of steps necessary for the name resolution of the FQDN of a blob/ADLS storage account to the private IP address of the corresponding private endpoint deployed in the hub virtual network:
+
+1. The virtual machine submits a DNS query for the FQDN of the `abc.blob.core.windows.net` blob storage account. The DNS request is sent to the Azure Firewall that acts as a DNS Proxy.
+2. Azure Firewall forwards the request to the default Azure DNS at `168.63.129.16`
+3. The authoritative query for `abc.blob.core.windows.net` returns `abc.privatelink.blob.core.windows.net` canonical name (CNAME)
+4. The DNS request for `abc.privatelink.blob.core.windows.net` is sent the `privatelink.blob.core.windows.net` Private DNS Zone that uses the A record for `abc` to return the associated private IP address
+5. The result is sent back to the Azure Firewall
+6. The result is sent back to the virtual machine
+7. The virtual machine uses the private IP address of the private endpoint to access the `abc.blob.core.windows.net` storage account
+
+## VM Custom Script Extemsion
+
+The two storage accounts are accessed via a different endpoint, hence their private endpoints need different private DNS zone:
+
+| Storage Service | Zone Name |
+| :-- | :--|
+| Blob service | privatelink.blob.core.windows.net |
+| Data Lake Storage Gen2 | privatelink.dfs.core.windows.net |
+
+For more information, see [Use private endpoints for Azure Storage](https://docs.microsoft.com/en-us/azure/storage/common/storage-private-endpoints#dns-changes-for-private-endpoints).
+
+The ARM template uses the [Azure Custom Script Extension](https://docs.microsoft.com/en-us/azure/virtual-machines/extensions/custom-script-linux) to download and run the following Bash script on the virtual machine. The script performs the following steps:
+
+- Validates the parameters received by the Custom Script extension
+- Updates the system and upgrades packages
+- Installs curl and traceroute packages
+- Install the NGINX web server on the local machine
+- Customize the main page of the NGINX web server to print the name of the local machine
+- Runs the nslookup command against the FQDN name of the two virtual machines
+- Runs curl against the web server of the virtual machines using their FQDN
+- Runs the nslookup command against the public URL of the ADLS Gen 2 storage account to verify that this gets resolved to a private address
+- Runs the nslookup command against the public URL of the second storage account to verify that this gets resolved to a private address
+- Downloads and installs the Azure CLI
+- Logins using the system-assigned managed identity of the virtual machine
+- Creates a file system in the ADLS Gen 2 storage account
+- Creates a directory in the file system
+- Creates a file in the directory with the content passed as a parameter
+
+```bash
```

Comment:
move the script code to a file in the sample, it will be easier to use and avoid a maintenance problem when the readme and script drift